File: app/src/main/java/com/nrs/nsnik/notes/data/FolderDao.java
/*
 * Copyright (C) 2017 nsnikhil
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
 */

package com.nrs.nsnik.notes.data;

import android.arch.lifecycle.LiveData;
import android.arch.persistence.room.Dao;
import android.arch.persistence.room.Delete;
import android.arch.persistence.room.Insert;
import android.arch.persistence.room.OnConflictStrategy;
import android.arch.persistence.room.Query;
import android.arch.persistence.room.Update;

import java.util.List;

@Dao
public interface FolderDao {

    @Query("SELECT * FROM FolderEntity")
    LiveData<List<FolderEntity>> getFoldersList();

    @Query("SELECT * FROM FolderEntity WHERE mUid = :id")
    LiveData<FolderEntity> getFolder(int id);

    @Query("SELECT * FROM FolderEntity WHERE mFolderName = :folderName")
    LiveData<FolderEntity> getFolderByName(String folderName);

    @Query("SELECT * FROM FolderEntity WHERE mFolderName LIKE :query")
    LiveData<List<FolderEntity>> getFolderByQuery(String query);

    @Query("SELECT * FROM FolderEntity WHERE mParentFolderName = :parentFolder")
    LiveData<List<FolderEntity>> getFolderByParent(String parentFolder);

    @Query("SELECT * FROM FolderEntity WHERE mParentFolderName = :parentFolder AND mIsPinned = 0 AND mIsLocked = 0")
    LiveData<List<FolderEntity>> getFolderByParentNoPinNoLock(String parentFolder);

    @Query("SELECT * FROM FolderEntity WHERE mParentFolderName = :parentFolder AND mIsPinned = 1 AND mIsLocked = 0")
    LiveData<List<FolderEntity>> getFolderByParentPinNoLock(String parentFolder);

    @Query("SELECT * FROM FolderEntity WHERE mParentFolderName = :parentFolder AND mIsPinned = 0 AND mIsLocked = 1")
    LiveData<List<FolderEntity>> getFolderByParentNoPinLock(String parentFolder);

    @Query("SELECT * FROM FolderEntity WHERE mParentFolderName = :parentFolder AND mIsPinned = 1 AND mIsLocked = 1")
    LiveData<List<FolderEntity>> getFolderByParentPinLock(String parentFolder);

    @Query("SELECT * FROM FolderEntity WHERE mIsPinned = :isPinned")
    LiveData<List<FolderEntity>> getFolderByPin(int isPinned);

    @Query("SELECT * FROM FolderEntity WHERE mIsLocked = :isLocked")
    LiveData<List<FolderEntity>> getFolderByLock(int isLocked);

    @Query("SELECT * FROM FolderEntity WHERE mColor = :color")
    LiveData<List<FolderEntity>> getFolderByColor(String color);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    long[] insertFolders(FolderEntity... folderEntities);

    @Delete
    void deleteFolders(FolderEntity... folderEntities);

    @Query("DELETE FROM FolderEntity WHERE mFolderName = :folderName")
    void deleteFolderByName(String folderName);

    @Query("DELETE FROM FolderEntity WHERE mParentFolderName = :parentFolderName")
    void deleteFolderByParent(String parentFolderName);

    @Update(onConflict = OnConflictStrategy.REPLACE)
    int updateFolders(FolderEntity... folderEntities);

}

File: app/src/main/java/com/nrs/nsnik/notes/data/FolderDao.kt
/*
 *     Credit Card Security V1  Copyright (C) 2018  sid-sun
 *     This program comes with ABSOLUTELY NO WARRANTY; for details type `show w'.
 *     This is free software, and you are welcome to redistribute it
 *     under certain conditions; type `show c' for details.
 *
 * The hypothetical commands `show w' and `show c' should show the appropriate
 * parts of the General Public License.  Of course, your program's commands
 * might be different; for a GUI interface, you would use an "about box".
 *
 *   You should also get your employer (if you work as a programmer) or school,
 * if any, to sign a "copyright disclaimer" for the program, if necessary.
 * For more information on this, and how to apply and follow the GNU GPL, see
 * <http://www.gnu.org/licenses/>.
 *
 *   The GNU General Public License does not permit incorporating your program
 * into proprietary programs.  If your program is a subroutine library, you
 * may consider it more useful to permit linking proprietary applications with
 * the library.  If this is what you want to do, use the GNU Lesser General
 * Public License instead of this License.  But first, please read
 * <http://www.gnu.org/philosophy/why-not-lgpl.html>.
 */

package com.nrs.nsnik.notes.data

import androidx.lifecycle.LiveData
import androidx.room.*

@Dao
interface FolderDao {

    @get:Query("SELECT * FROM FolderEntity")
    val foldersList: LiveData<List<FolderEntity>>

    @Query("SELECT * FROM FolderEntity")
    fun getFolders(): LiveData<List<FolderEntity>>

    @Query("SELECT * FROM FolderEntity WHERE uid = :id")
    fun getFolder(id: Int): LiveData<FolderEntity>

    @Query("SELECT * FROM FolderEntity WHERE folderName = :folderName")
    fun getFolderByName(folderName: String): LiveData<FolderEntity>

    @Query("SELECT * FROM FolderEntity WHERE folderName LIKE :query|| '%'")
    fun getFolderByQuery(query: String): LiveData<List<FolderEntity>>

    @Query("SELECT * FROM FolderEntity WHERE parentFolderName = :parentFolder")
    fun getFolderByParent(parentFolder: String): LiveData<List<FolderEntity>>

    @Query("SELECT * FROM FolderEntity WHERE parentFolderName = :parentFolder AND pinned = 0 AND locked = 0")
    fun getFolderByParentNoPinNoLock(parentFolder: String): LiveData<List<FolderEntity>>

    @Query("SELECT * FROM FolderEntity WHERE parentFolderName = :parentFolder AND pinned = 1 AND locked = 0")
    fun getFolderByParentPinNoLock(parentFolder: String): LiveData<List<FolderEntity>>

    @Query("SELECT * FROM FolderEntity WHERE parentFolderName = :parentFolder AND pinned = 0 AND locked = 1")
    fun getFolderByParentNoPinLock(parentFolder: String): LiveData<List<FolderEntity>>

    @Query("SELECT * FROM FolderEntity WHERE parentFolderName = :parentFolder AND pinned = 1 AND locked = 1")
    fun getFolderByParentPinLock(parentFolder: String): LiveData<List<FolderEntity>>

    @Query(" SELECT * FROM FolderEntity WHERE parentFolderName = :parentFolder ORDER BY pinned DESC")
    fun getFolderByParentOrdered(parentFolder: String): LiveData<List<FolderEntity>>

    @Query("SELECT * FROM FolderEntity WHERE pinned = :isPinned")
    fun getFolderByPin(isPinned: Int): LiveData<List<FolderEntity>>

    @Query("SELECT * FROM FolderEntity WHERE locked = :isLocked")
    fun getFolderByLock(isLocked: Int): LiveData<List<FolderEntity>>

    @Query("SELECT * FROM FolderEntity WHERE color = :color")
    fun getFolderByColor(color: String): LiveData<List<FolderEntity>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun insertFolders(vararg folderEntities: FolderEntity): LongArray

    @Delete
    fun deleteFolders(vararg folderEntities: FolderEntity)

    @Query("DELETE FROM FolderEntity WHERE folderName = :folderName")
    fun deleteFolderByName(folderName: String)

    @Query("DELETE FROM FolderEntity WHERE parentFolderName = :parentFolderName")
    fun deleteFolderByParent(parentFolderName: String)

    @Update(onConflict = OnConflictStrategy.REPLACE)
    fun updateFolders(vararg folderEntities: FolderEntity): Int

    @Query("UPDATE FolderEntity SET pinned = :pin WHERE uid = :id")
    fun changeFolderPinStatus(id: Int, pin: Int)

    @Query("UPDATE FolderEntity SET locked = :lock WHERE uid = :id")
    fun changeFolderLockStatus(id: Int, lock: Int)

    @Query("UPDATE FolderEntity SET parentFolderName = :parentFolderName WHERE uid = :id")
    fun changeFolderParent(id: Int, parentFolderName: String)

}
